implementations/kmp-json-schema-validator/src/main/kotlin/BowtieSampsonSchemaValidatorLauncher.kt

```diff
@@ -0,0 +1,214 @@
+            ignoreUnknownKeys = true
+            prettyPrint = false
+            encodeDefaults = true
+        }
+        val processor = BowtieSampsonSchemaValidatorLauncher(outputWriter, json)
+        input.lines().forEach {
+            val command = json.decodeFromString(Command.serializer(), it)
+            processor.handle(command)
+        }
+    } finally {
+        outputWriter.flush()
+    }
+}
+
+private val SUPPORTED_DIALECTS: Set<String> = hashSetOf(
+    "http://json-schema.org/draft-07/schema#",
+)
+
+// TODO: remove when remote ref support is added
+/**
+ * All these cases are ignored because they contain remote refs
+ * Library does not support them yet.
+ */
+private val IGNORED_CASES: Set<String> = hashSetOf(
+    "validate definition against metaschema",
+    "base URI change - change folder",
+    "base URI change - change folder in subschema",
+    "base URI change",
+    "retrieved nested refs resolve relative to their URI not \$id",
+    "\$ref to \$ref finds location-independent \$id"
+)
+
+enum class Result { CONTINUE, STOP }
+
+class BowtieSampsonSchemaValidatorLauncher(
+    private val writer: BufferedWriter,
+    private val json: Json,
+) {
+    private var started: Boolean = false
+    private var currentDialect: String = ""
+    private val libraryName: String
+    private val libraryVersion: String
+    private val libraryHomepage: String
+    private val libraryIssues: String
+
+    init {
+        val attributes: Attributes = javaClass.getResourceAsStream("META-INF/MANIFEST.MF").use {
+            requireNotNull(it) { "cannot find manifest file" }
+            Manifest(it).mainAttributes
+        }
+        with(attributes) {
+            libraryName = getValue("Implementation-Name")
+            libraryVersion = getValue("Implementation-Version")
+            libraryHomepage = getValue("Implementation-Homepage")
+            libraryIssues = getValue("Implementation-Issues")
+        }
+    }
+
+    fun handle(command: Command): Result {
+        return when (command) {
+            is Command.Start -> handleStart(command)
+            is Command.Dialect -> handleDialect(command)
+            is Command.Run -> {
+                handleRun(command)
+                Result.CONTINUE
+            }
+
+            Command.Stop -> Result.STOP
+        }
+    }
+
+    private fun shouldSkipCase(caseDescription: String): String? {
+        return when {
+            caseDescription.endsWith(" format") -> "impl does not support format assertion"
+            caseDescription in IGNORED_CASES || caseDescription.contains("remote ref") ->
+                "impl does not support remote schema loading yet"
+            else -> null
+        }
+    }
+
+    private fun shouldSkipTest(caseDescription: String, testDescription: String): Boolean {
+        // TODO: check tests to skip
+        return false
+    }
+
+    private fun handleRun(command: Command.Run) {
+        requireStarted()
+        requireDialect()
```

Comment:
I think @harrel56 already pointed out some of the slightly trickiness here, but this line / this check too is not necessarily needed/true -- if your harness is sent the schema `{"$schema": "http://json-schema.org/draft-07/schema#", "type": "integer"}` no dialect request is needed/required, the appropriate dialect is already indicated in the schema.

The dialect request is essentially needed only for schemas that don't have a `$schema` property.

Reply:
Right!